File: PHP/max_number.php
<?php # max_number.php - Nate Nasteff 2020

/*
** This function will take an integer argument and
** rearrange the digits to produce the highest possible
** number. Once split into a string array, the numbers are sorted,
** then iterated over, adding each value multiplied by the digit's
** poisition left of the decimal point.
** IE (int=417 => sorted_arr=[1,4,7] => 1*1 + 4*10 + 7*100 = 741)
*/

$max_number = function ($num) {
  $max = 0;
  $multi = 1;
  $arr = array_map('intval', str_split($num));
  sort($arr);
  for ($i = 0; $i < count($arr); $i++) {
    // This prevents int out of bounds
    if (count($arr) > 9)
      return "Number too big!";
    $max += $arr[$i] * $multi;
    $multi *= 10;
  }
  return $max;
};

$int = 256783441;
echo "\nOriginal number:\n$int\n\nMax number:\n"
  . $max_number($int) . "\n";


/*
** Example I/O
**
** INPUT:
** 256783441
**
** OUTPUT:
**
** Original number:
** 256783441
**
** Max number:
** 87654432 
** 
** Time complexity: O(n)
** Space complexity: O(1)
** 
*/

File: Java/Algorithms/MaxNumber.java
package Java.Algorithms;

/**
 ** @author Nathan Nasteff
 */

public class MaxNumber {

  /*
   * This function takes an integer as an argument and returns the highest
   * possible number combination that can be made from the original integer.
   */

  public static int getMax(int num) {

    int arr[] = new int[10];
    String str = Integer.toString(num);
    for (int i = 0; i < str.length(); i++) {
      arr[str.charAt(i) - '0']++;
    }

    int max = 0, multiplier = 1;

    for (int i = 0; i <= 9; i++) {
      while (arr[i] > 0) {
        max = max + (i * multiplier);
        arr[i]--;
        multiplier = multiplier * 10;
      }
    }

    return max;
  }

  public static void main(String[] args) {
    int a = 120345676;
    System.out.println("Original number:\n" + a + "\n");
    a = getMax(a);
    System.out.println("Largest number possible:\n" + a + "\n");
  }
}

/*
 ** Sample I/O
 **
 ** INPUT:
 ** 
 ** 120345676
 **
 ** OUTPUT:
 **
 ** Original number: 120345676
 **
 ** Largest number possible: 766543210
 **
 **
 ** Time complexity - O(n)
 **
 ** Space complexity - O(n)
 **
 */
